common/processes/platform/tiny_process_library/tiny_process_library.cpp

```diff
@@ -192,31 +201,39 @@ void Process::SetupAsyncWait() {
 
 		if (async_wait_data->handler) {
 			async_wait_data_->event_loop->Post([async_wait_data, this]() {
-				// This inner function is executed on the event loop, and here the
-				// object may have been either cancelled or destroyed before we get
-				// here. By having our own copy of the shared pointer, it survives
-				// destruction, and we can test whether we should still proceed
-				// here.
-				unique_lock lock(async_wait_data->data_mutex);
-
-				if (async_wait_data->handler) {
-					auto handler = async_wait_data->handler;
-
-					async_wait_data->event_loop = nullptr;
-					async_wait_data->handler = nullptr;
-
-					// Unlock in case the handler calls back into this object.
-					lock.unlock();
-					auto status = GetExitStatus();
-					handler(ErrorBasedOnExitStatus(status));
-				}
+				Process::AsyncWaitInternalHandler(async_wait_data, *this);
 			});
 		}
 
+		async_wait_data->process_ended = true;
+
 		return status;
 	});
 }
 
+void Process::AsyncWaitInternalHandler(shared_ptr<AsyncWaitData> async_wait_data, Process &proc) {
```

Comment:
Should this even be a member function? It doesn't work with any private members and in practice it actually takes `*this` as the second argument (which is a bit confusing and deserves a comment, actually).

Reply:
It was meant to make it clearer that our own `async_wait_data` must be used over the one in the member variables. But you're right, it's not clearer, so I'll just change it to a member function and explain `async_wait_data` separately.